### ISSUE TITLE
If no safe area, TextPainter.getPositionForOffset reports incorrect multiline positions for mobile

### ISSUE THREAD

**Issue by stuz5000** [flutter/flutter#45020]
In the following, clicking on letter:
- in the first row should print a number between 0..9
- in the second row should print a number between 10..19
- in the third row should print a number between 20..29

You only see correct behavior if SafeArea is uncommented.
Without the SafeArea, text positions are incorrectly report.

Possibly related to: https://github.com/flutter/flutter/issues/44121

```dart
import 'package:flutter/material.dart';
final loremIpsum = "012345678\n"
    "012345678\n"
    "012345678\n"
    "012345678\n"
    "012345678\n";

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Welcome to Flutter',
        home: Material(
//            child: SafeArea(
          child: Center(
            child: Container(
                width: 350,
                color: Color.fromARGB(100, 0, 0, 0),
                child: ListView.separated(
                    itemCount: 3,
                    separatorBuilder: (_, __) => Divider(),
                    itemBuilder: (_, __) => Card(child: SelectText()))),
          ),
//        )
      )
    );
  }
}

class SelectText extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    //    return redbox();
    // Set width to max allowed by parent
    return LayoutBuilder(
        builder: (BuildContext context, BoxConstraints constraints) {
      final textPainter = TextPainter(
        text: TextSpan(
          text: loremIpsum,
          style: TextStyle(height: 2.5),
        ),
        textDirection: TextDirection.ltr,
      );

      final width = constraints.maxWidth;
      textPainter.layout(
        minWidth: 20,
        maxWidth: width,
      );
      final height = textPainter.height;

      return Container(
          width: width,
          height: height,
          color: Colors.grey,
          child: GestureDetector(
            onTapDown: (details) {
              // BUG. On web- position is only correct for first line.
              print(
                  "-> ${textPainter.getPositionForOffset(details.localPosition).offset}");
            },
            child: CustomPaint(
              size: Size(width, height), // Parent width, text height
              painter: TextCustomPainter(textPainter),
            ),
          ));
    });
  }
}

class TextCustomPainter extends CustomPainter {
  TextPainter textPainter;

  TextCustomPainter(this.textPainter, {Listenable repaint})
      : super(repaint: repaint);

  @override
  void paint(Canvas canvas, Size size) {
    textPainter.paint(canvas, Offset(0, 0));
  }

  @override
  bool shouldRepaint(CustomPainter old) {
    return false;
  }
}

```
Target OS version/browser: Nexus 7 API 29, developing in ubuntu 18.

## Flutter doctor

```
[✓] Flutter (Channel master, v1.10.16-pre.68, on Linux, locale en_US.UTF-8)
    • Flutter version 1.10.16-pre.68 at /home/sir/development/flutter
    • Framework revision cb614d1339 (8 days ago), 2019-11-07 08:24:19 -0800
    • Engine revision c448d39d53
    • Dart version 2.7.0

[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /home/sir/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /home/sir/soft/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /home/sir/soft/android-studio
    • Flutter plugin version 41.1.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] IntelliJ IDEA Ultimate Edition (version 2019.2)
    • IntelliJ at /snap/intellij-idea-ultimate/181
    • Flutter plugin version 41.0.4
    • Dart plugin version 192.7402

[✓] Connected device (3 available)
    • Android SDK built for x86 • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • Chrome                    • chrome        • web-javascript • Google Chrome 76.0.3809.132
    • Web Server                • web-server    • web-javascript • Flutter Tools

```

**Comment by stuz5000** [comment]
If a SafeArea is present, but ListView:padding is specified, it again failed to report the correct text position.

```
import 'package:flutter/material.dart';
final loremIpsum = "012345678\n"
    "012345678\n"
    "012345678\n"
    "012345678\n"
    "012345678\n";

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Welcome to Flutter',
        home: Material(
            child: SafeArea( //  <<<< Broken if missing
          child: Center(
            child: Container(
                width: 350,
                color: Color.fromARGB(100, 0, 0, 0),
                child: ListView.separated(
                    itemCount: 3,
                    padding: const EdgeInsets.all(20.0),  //  <<<<  Broken if added
                    separatorBuilder: (_, __) => Divider(),
                    itemBuilder: (_, __) => Card(child: SelectText()))),
          ),
        )
      )
    );
  }
}

class SelectText extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    //    return redbox();
    // Set width to max allowed by parent
    return LayoutBuilder(
        builder: (BuildContext context, BoxConstraints constraints) {
      final textPainter = TextPainter(
        text: TextSpan(
          text: loremIpsum,
          style: TextStyle(height: 2.5),
        ),
        textDirection: TextDirection.ltr,
      );

      final width = constraints.maxWidth;
      textPainter.layout(
        minWidth: 20,
        maxWidth: width,
      );
      final height = textPainter.height;

      return Container(
          width: width,
          height: height,
          color: Colors.grey,
          child: GestureDetector(
            onTapDown: (details) {
              // BUG. On web- position is only correct for first line.
              print(
                  "-> ${textPainter.getPositionForOffset(details.localPosition).offset}");
            },
            child: CustomPaint(
              size: Size(width, height), // Parent width, text height
              painter: TextCustomPainter(textPainter),
            ),
          ));
    });
  }
}

class TextCustomPainter extends CustomPainter {
  TextPainter textPainter;

  TextCustomPainter(this.textPainter, {Listenable repaint})
      : super(repaint: repaint);

  @override
  void paint(Canvas canvas, Size size) {
    textPainter.paint(canvas, Offset(0, 0));
  }

  @override
  bool shouldRepaint(CustomPainter old) {
    return false;
  }
}
```

**Comment by goderbauer** [comment]
@stuz5000 Is this happening when you run a flutter app natively on mobile? Or are you using Flutter Web?

**Comment by stuz5000** [comment]
This happen within the Nexus 7 2012 API 29 Emulator.

**Comment by stuz5000** [comment]
Any word on whether this will be resolved. Only being able to deal with single-line text is blocking our app release. - Thanks

**Comment by stuz5000** [comment]
Looks like is related to #33523

**Comment by markusaksli-nc** [comment]
Closing since this seems fixed on the latest master `1.22.0-2.0.pre.78`.
<details>
 <summary>flutter doctor -v</summary>

```
[√] Flutter (Channel master, 1.22.0-2.0.pre.78, on Microsoft Windows [Version 10.0.19041.450], locale en-US)
    • Flutter version 1.22.0-2.0.pre.78 at C:\Development\flutter_master
    • Framework revision d0596582fe (2 hours ago), 2020-08-25 23:26:06 -0700
    • Engine revision b1d9f863db
    • Dart version 2.10.0 (build 2.10.0-56.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at C:\Users\marku\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.6.5)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.6.30320.27
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.0)
    • Android Studio at C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] Connected device (5 available)
    • SM G950F (mobile)           • ce12171c51cc001c03 • android-arm64  • Android 9 (API 28)
    • sdk gphone x86 arm (mobile) • emulator-5554      • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)           • windows            • windows-x64    • Microsoft Windows [Version 10.0.19041.450]
    • Web Server (web)            • web-server         • web-javascript • Flutter Tools
    • Chrome (web)                • chrome             • web-javascript • Google Chrome 84.0.4147.135

• No issues found!
```
</details>
